src/Contract/Test/Assert.purs

```diff
@@ -0,0 +1,631 @@
+  wrappedContract = foldr wrapAssertion contract assertions
+
+-- | Accepts an array of checks and interprets them into a `Contract`.
+runChecks
+  :: forall (a :: Type)
+   . Array (ContractCheck a)
+  -> ContractAssertion a
+  -> Contract a
+runChecks assertions contract = do
+  eiResult /\ failures <- collectAssertionFailures assertions contract
+  if null failures then either
+    (liftEffect <<< throwError <<< error <<< reportException)
+    pure
+    eiResult
+  else do
+    let
+      errorStr = either reportException (const "") eiResult
+      errorReport =
+        String.trim
+          ( errorStr <> "\n\n" <>
+              printContractAssertionFailures (Array.fromFoldable failures)
+          ) <> "\n"
+    -- error trace from the exception itself will be appended here
+    liftEffect $ throwError $ error errorReport
+  where
+  reportException :: Error -> String
+  reportException error = "\n\nAn exception has been thrown: \n\n" <> show error
+
+tellFailure
+  :: ContractAssertionFailure -> ContractAssertion Unit
+tellFailure failure = do
+  ask >>= liftEffect <<< Ref.modify_ (Cons failure)
+
+checkNewUtxosAtAddress
+  :: forall (a :: Type)
+   . Labeled Address
+  -> TransactionHash
+  -> (Array TransactionOutputWithRefScript -> ContractAssertion a)
+  -> ContractAssertion a
+checkNewUtxosAtAddress addr txHash check =
+  lift (utxosAt $ unlabel addr) >>= \utxos ->
+    check $ Array.fromFoldable $ Map.values $
+      Map.filterKeys (\oref -> (unwrap oref).transactionId == txHash) utxos
+
+-- | Sets a limit on `ExUnits` budget. All ExUnits values of all submitted transactions are combined. Transactions that are constructed, but not submitted, are not considered.
+-- | The execution of the `Contract` will not be interrupted in case the `ExUnits` limit is reached.
+checkExUnitsNotExceed
+  :: forall (a :: Type)
+   . ExUnits
+  -> ContractCheck a
+checkExUnitsNotExceed maxExUnits contract = do
+  (ref :: Ref ExUnits) <- liftEffect $ Ref.new { mem: zero, steps: zero }
+  let
+    submitHook :: Transaction -> Effect Unit
+    submitHook tx = do
+      let
+        (newExUnits :: ExUnits) = sum $ tx ^..
+          _witnessSet
+            <<< _redeemers
+            <<< non []
+            <<< traversed
+            <<< to (unwrap >>> _.exUnits)
+      Ref.modify_ (add newExUnits) ref
+
+    setSubmitHook :: ContractEnv -> ContractEnv
+    setSubmitHook =
+      prop (Proxy :: Proxy "hooks") <<< prop (Proxy :: Proxy "onSubmit")
+        -- Extend a hook action if it exists, or set it to `Just submitHook`
+        %~ maybe (Just submitHook)
+          \oldHook -> Just \tx -> do
+            -- ignore possible exception from the old hook
+            void $ try $ oldHook tx
+            submitHook tx
+
+    finalize :: ContractAssertion Unit
+    finalize = do
+      exUnits <- liftEffect $ Ref.read ref
+      assertContract (MaxExUnitsExceeded (ExpectedActual maxExUnits exUnits))
+        (maxExUnits >= exUnits)
```

Comment:
This is a bug, we must compare the fields independently. For eg, the ord instance on records does the following:

```
> { a: 3, b: 1000 } > { a: 10, b: 20 }  
> false
```

Reply:
Thank you!